From: Niko Tyni <ntyni@debian.org>
Date: Mon, 16 Jul 2018 10:12:37 +0300
Subject: Disable debugperl stack extension checks for binary compatibility
 with perl

When an XS module is built without -DDEBUGGING but the interpreter is,
the debugging stack high-water mark checking on the interpreter side
doesn't work properly. This makes /usr/bin/debugperl less useful than
it could be as it can no longer use many XS module packages.

Patch the check away for now. We hope upstream will provide a Configure
option for this in the future.

Bug: https://github.com/Perl/perl5/issues/16607
Bug-Debian: https://bugs.debian.org/902779
---
 dump.c   | 3 +++
 pp_hot.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/dump.c b/dump.c
index f03c3f6..737be76 100644
--- a/dump.c
+++ b/dump.c
@@ -2535,6 +2535,8 @@ Perl_runops_debug(pTHX)
         ++PL_op_exec_cnt[PL_op->op_type];
 #endif
 #if defined DEBUGGING && !defined DEBUGGING_RE_ONLY
+/* disabled on Debian due to https://bugs.debian.org/902779 */
+#ifndef DEBIAN
         if (PL_curstackinfo->si_stack_hwm < PL_stack_sp - PL_stack_base)
             Perl_croak_nocontext(
                 "panic: previous op failed to extend arg stack: "
@@ -2542,6 +2544,7 @@ Perl_runops_debug(pTHX)
                     PL_stack_base, PL_stack_sp,
                     PL_stack_base + PL_curstackinfo->si_stack_hwm);
         PL_curstackinfo->si_stack_hwm = PL_stack_sp - PL_stack_base;
+#endif
 #endif
 	if (PL_debug) {
             ENTER;
diff --git a/pp_hot.c b/pp_hot.c
index 95f6d64..4f29316 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -5277,6 +5277,8 @@ PP(pp_entersub)
 	CvXSUB(cv)(aTHX_ cv);
 
 #if defined DEBUGGING && !defined DEBUGGING_RE_ONLY
+/* disabled on Debian due to https://bugs.debian.org/902779 */
+#ifndef DEBIAN
         /* This duplicates the check done in runops_debug(), but provides more
          * information in the common case of the fault being with an XSUB.
          *
@@ -5290,6 +5292,7 @@ PP(pp_entersub)
                     HvNAME(GvSTASH(CvGV(cv))), GvNAME(CvGV(cv)), CvFILE(cv),
                     PL_stack_base, PL_stack_sp,
                     PL_stack_base + PL_curstackinfo->si_stack_hwm);
+#endif
 #endif
 	/* Enforce some sanity in scalar context. */
 	if (is_scalar) {
